sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,777 @@
+            unsafe {
+                sol_bignum_mod_exp(
+                    &mut *bignum_ptr as *mut _ as *mut u64,
+                    &self.0 as *const _ as *const u64,
+                    &exponent.0 as *const _ as *const u64,
+                    &modulus.0 as *const _ as *const u64,
+                );
+            }
+            Self(*bignum_ptr)
+        }
+    }
+    /// Hashed Generator
+    pub fn hashed_generator(&self, a: &BigNumber, n: &BigNumber, nonce: &[u8]) -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            use hkdf::Hkdf;
+            use openssl::bn::{BigNum, BigNumContext};
+
+            let u_bn = unsafe { &*(self.0 as *mut BigNum) };
+            let a_bn = unsafe { &*(a.0 as *mut BigNum) };
+            let n_bn = unsafe { &*(n.0 as *mut BigNum) };
+            println!("u:{}", u_bn);
+            println!("a:{}", a_bn);
+            println!("n:{}", n_bn);
+            // hash_generator
+            let mut transcript = u_bn.as_ref().to_vec();
+            transcript.append(&mut a_bn.as_ref().to_vec());
+            transcript.extend_from_slice(nonce);
+            println!("t:{:?}", transcript);
+            let length = n_bn.num_bytes();
+            println!("len:{}", length);
+            let h = Hkdf::<blake3::Hasher>::new(
+                Some(b"SST_SALT_HASH_TO_GENERATOR_"),
+                transcript.as_slice().as_ref(),
+            );
+            let mut okm = vec![0u8; length as usize];
+            h.expand(b"", &mut okm).unwrap();
+            println!("okm:{:?}", okm);
+            let okm_bn = BigNum::from_slice(okm.as_slice()).unwrap();
+
+            let mut res_bn = Box::new(BigNum::new().unwrap());
+            res_bn
+                .mod_sqr(
+                    okm_bn.as_ref(),
+                    n_bn.as_ref(),
+                    &mut BigNumContext::new().unwrap(),
+                )
+                .unwrap();
+            let rwptr = Box::into_raw(res_bn);
+            Self(rwptr as u64)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_hashed_generator(
+                    self_ptr: *const u64,
+                    a: *const u64,
+                    n: *const u64,
+                    nonce: *const u64,
+                    hg_ptr: *mut u64,
+                ) -> u64;
+            }
+            let mut bignum_ptr = nonce.len() as u64;
+            unsafe {
+                sol_bignum_hashed_generator(
+                    &self.0 as *const _ as *const u64,
+                    &a.0 as *const _ as *const u64,
+                    &n.0 as *const _ as *const u64,
+                    nonce as *const _ as *const u64,
+                    &mut bignum_ptr as *mut _ as *mut u64,
+                );
+            }
+            Self(bignum_ptr)
+        }
+    }
+
+    /// Hash to Prime
+    pub fn hash_to_prime(u: &BigNumber, a: &BigNumber, z: &BigNumber, nonce: &[u8]) -> Self {
```

Comment:
Consider moving hash operation to a hash.rs module.

Reply:
Removed `hash_to_prime` from PR... see 01ed0b9d229749506e0065fed170cac1310741b1